### ISSUE TITLE
Reproducibility of SimCLR

### ISSUE THREAD

**Issue by asparius** [lightly-ai/lightly#1419]
I could not reproduce the results of SimCLR in CIFAR10.  The entire code and hyperparameters are below.

num_workers = 4
batch_size = 512
seed = 1
max_epochs_pretrain = 1000
max_epochs_linear= 100
input_size = 32
num_ftrs = 32

path_to_train = "cifar10/cifar10/train/"
path_to_test = "cifar10/cifar10/test/"

transform = SimCLRTransform(input_size=input_size, vf_prob=0.5, rr_prob=0.5)

train_classifier_transforms = torchvision.transforms.Compose(
    [
        torchvision.transforms.RandomCrop(32, padding=4),
        torchvision.transforms.RandomHorizontalFlip(),
        torchvision.transforms.ToTensor(),
        torchvision.transforms.Normalize(
            mean=utils.IMAGENET_NORMALIZE["mean"],
            std=utils.IMAGENET_NORMALIZE["std"],
        ),
    ]
)

test_transforms = torchvision.transforms.Compose(
    [
        torchvision.transforms.Resize((32, 32)),
        torchvision.transforms.ToTensor(),
        torchvision.transforms.Normalize(
            mean=utils.IMAGENET_NORMALIZE["mean"],
            std=utils.IMAGENET_NORMALIZE["std"],
        ),
    ]
)

dataset_train_simclr = LightlyDataset(input_dir=path_to_train, transform=transform)

dataset_train_classifier = LightlyDataset(
    input_dir=path_to_train, transform=train_classifier_transforms
)

dataset_test = LightlyDataset(input_dir=path_to_test, transform=test_transforms)

dataloader_train_simclr = torch.utils.data.DataLoader(
    dataset_train_simclr,
    batch_size=batch_size,
    shuffle=True,
    drop_last=True,
    num_workers=num_workers,
)

dataloader_train_classifier = torch.utils.data.DataLoader(
    dataset_train_classifier,
    batch_size=batch_size,
    shuffle=True,
    drop_last=True,
    num_workers=num_workers,
)

dataloader_test = torch.utils.data.DataLoader(
    dataset_test,
    batch_size=batch_size,
    shuffle=False,
    drop_last=False,
    num_workers=num_workers,
)

from lightly.loss import NTXentLoss
from lightly.models.modules.heads import SimCLRProjectionHead
from lightly.models.utils import (
    deactivate_requires_grad
)

class SimCLRModel(pl.LightningModule):
    def __init__(self):
        super().__init__()

        resnet = torchvision.models.resnet18()
        self.backbone = nn.Sequential(*list(resnet.children())[:-1])

        hidden_dim = resnet.fc.in_features
        self.projection_head = SimCLRProjectionHead(hidden_dim, hidden_dim, 128)

        self.criterion = NTXentLoss()

    def forward(self, x):
        h = self.backbone(x).flatten(start_dim=1)
        z = self.projection_head(h)
        return z

    def training_step(self, batch, batch_idx):
        (x0, x1), _, _ = batch
        z0 = self.forward(x0)
        z1 = self.forward(x1)
        loss = self.criterion(z0, z1)
        self.log("train_loss_ssl", loss)
        return loss

    def configure_optimizers(self):
        optim = torch.optim.SGD(
            self.parameters(), lr=6e-1, momentum=0.9, weight_decay=1e-6,nesterov=True
        )
        scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optim, max_epochs_pretrain)
        return [optim], [scheduler]

class Classifier(pl.LightningModule):
    def __init__(self, backbone):
        super().__init__()
        
        self.backbone = backbone

        
        deactivate_requires_grad(backbone)

        
        self.fc = nn.Linear(512, 10)

        self.criterion = nn.CrossEntropyLoss()
        self.validation_step_outputs = []

    def forward(self, x):
        y_hat = self.backbone(x).flatten(start_dim=1)
        y_hat = self.fc(y_hat)
        return y_hat

    def training_step(self, batch, batch_idx):
        x, y, _ = batch
        y_hat = self.forward(x)
        loss = self.criterion(y_hat, y)
        self.log("train_loss_fc", loss)
        return loss

    

    def validation_step(self, batch, batch_idx):
        x, y, _ = batch
        y_hat = self.forward(x)
        y_hat = torch.nn.functional.softmax(y_hat, dim=1)

       
        _, predicted = torch.max(y_hat, 1)
        num = predicted.shape[0]
        correct = (predicted == y).float().sum()
        self.validation_step_outputs.append((num, correct))
        return num, correct

    def on_validation_epoch_end(self):
        
        if self.validation_step_outputs:
            total_num = 0
            total_correct = 0
            for num, correct in self.validation_step_outputs:
                total_num += num
                total_correct += correct
            acc = total_correct / total_num
            self.log("val_acc", acc, on_epoch=True, prog_bar=True)
            self.validation_step_outputs.clear()

    def configure_optimizers(self):
        optim = torch.optim.SGD(self.fc.parameters(), lr=0.2,momentum=0.9,nesterov=True)
        scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optim, max_epochs_linear)
        return [optim], [scheduler]

wandb_logger_train = pl.loggers.WandbLogger(name="pretraining",log_model="all")

model = SimCLRModel()
trainer = pl.Trainer(max_epochs=max_epochs_pretrain, devices=1, accelerator="gpu",logger=wandb_logger_train)
trainer.fit(model, dataloader_train_simclr)

wandb_logger = pl.loggers.WandbLogger(name="finetuning",log_model="all")
model.eval()
classifier = Classifier(model.backbone)
trainer = pl.Trainer(max_epochs=max_epochs_linear, devices=1, accelerator="gpu",logger=wandb_logger)
trainer.fit(classifier, dataloader_train_classifier, dataloader_test)

**Comment by guarin** [comment]
Hi! We use KNN evaluation for Cifar10, from your code it looks like you are finetuning the model instead. The code for the benchmark results in the [docs](https://docs.lightly.ai/self-supervised-learning/getting_started/benchmarks.html#cifar-10) can be found at: https://github.com/lightly-ai/lightly/blob/master/docs/source/getting_started/benchmarks/cifar10_benchmark.py

**Comment by asparius** [comment]
You are right but KNN is not a common evaluation setting in SSL methods. Linear evaluation scores will be better if you include them as well.

**Comment by guarin** [comment]
Our ImageNet benchmarks use KNN, linear, and finetune evaluation: https://docs.lightly.ai/self-supervised-learning/getting_started/benchmarks.html#imagenet

We sadly didn't have time yet to convert the old Cifar10 benchmarks to the new format.

**Comment by asparius** [comment]
Why optimizer hyperparameters are different from the official SimCLR repository, i.e lr, weight decay?

**Comment by asparius** [comment]
> Why optimizer hyperparameters are different from the official SimCLR repository, i.e lr, weight decay?

Cant get more than 0.765 with SimCLR

**Comment by guarin** [comment]
> Why optimizer hyperparameters are different from the official SimCLR repository, i.e lr, weight decay?

I don't recall exactly why they are not the same. Did you also try with the original settings: https://github.com/google-research/simclr#pretraining ?

**Comment by IgorSusmelj** [comment]
Models like SimCLR use different hyperparams and model backbones for cifar10 as the images are smaller and the dataset differs quite a bit from datasets like ImageNet. I think there is an appendix or so that shows the exact hyperparams used for the original SimCLR paper. The hyperparams we use for ImageNet should match 100% the ones from the paper.

**Comment by asparius** [comment]
I have found the problem. In CIFAR10 experiments, the first max pooling is discarded and the first conv layer kernel size is (3,3) and with stride(1,1) instead of (7,7) and (3,3). I get around 87 percent with resnet18 and I have also skipped this finetune after 4th block thing which will make it closer to the paper's results. You might want to update your benchmarks accordingly.
